src/repository.ts

```diff
@@ -0,0 +1,393 @@
+
+    const key = Buffer.from(env.getString('SALESFORCE_KEY'), 'base64').toString();
+    this.keyPath = path.join(os.tmpdir(), 'salesforce-cli.key');
+    await fs.writeFile(this.keyPath, key);
+    this.logger.debug(`Wrote key to ${this.keyPath}`);
+  }
+
+  public async sign(target?: string): Promise<SigningResponse> {
+    const repsonse = packAndSignApi.doPackAndSign(
+      {
+        signatureurl: Signer.SIGNATURE_URL,
+        publickeyurl: Signer.PUBLIC_KEY_URL,
+        privatekeypath: this.keyPath,
+        target,
+      },
+      this.ux
+    );
+    await fs.unlink(this.keyPath);
+    return repsonse;
+  }
+
+  protected async init(): Promise<void> {
+    this.logger = await Logger.child(this.constructor.name);
+    await this.prepare();
+  }
+}
+
+abstract class Repository extends AsyncOptionalCreatable {
+  protected ux: UX;
+  private stepCounter = 1;
+
+  public constructor(ux: UX) {
+    super(ux);
+    this.ux = ux;
+  }
+
+  public install(silent = false): void {
+    this.execCommand('yarn install', silent);
+  }
+
+  public build(silent = false): void {
+    this.execCommand('yarn build', silent);
+  }
+
+  public pushChangesToGit(): void {
+    const currentBranch = exec('npx git branch --show-current', { silent: true }).stdout;
+    const cmd = `npx git push --no-verify --follow-tags origin ${currentBranch}`;
+    this.ux.log(cmd);
+    exec(cmd, { silent: false });
+  }
+
+  public revertUnstagedChanges(): void {
+    const changedFiles = exec('npx git diff --name-only', { silent: true })
+      .stdout.split(os.EOL)
+      .filter((f) => !!f);
+    changedFiles.forEach((file) => {
+      exec(`git checkout -- ${file}`, { silent: false });
+    });
+  }
+
+  public printStage(msg: string): void {
+    this.ux.log(chalk.green.bold(`${os.EOL}${this.stepCounter}) ${msg}`));
+    this.stepCounter += 1;
+  }
+
+  public async uploadSignature(signResult: SigningResponse): Promise<void> {
+    const result = await upload(signResult.filename, 'dfc-data-production', 'media/salesforce-cli/signatures');
+    this.ux.log(`ETag: ${result.ETag}`);
+    this.ux.log(`VersionId: ${result.VersionId}`);
+  }
+
+  protected execCommand(cmd: string, silent?: boolean): ShellString {
+    if (!silent) this.ux.log(`${chalk.dim(cmd)}${os.EOL}`);
+    const result = exec(cmd, { silent });
+    if (result.code !== 0) {
+      throw new SfdxError(result.stderr, 'FailedCommandExecution');
+    } else {
+      return result;
+    }
+  }
+
+  public abstract getSuccessMessage(): string;
+  public abstract validate(): VersionValidation | VersionValidation[];
+  public abstract publish(options: PublishOpts): void;
+  public abstract verifySignature(packageNames?: string[]): void;
+  public abstract async sign(packageNames?: string[]): Promise<SigningResponse | SigningResponse[]>;
+  public abstract async waitForAvailability(): Promise<boolean>;
+  protected abstract async init(): Promise<void>;
+}
+
+export class MultiPackageRepo extends Repository {
```

Comment:
Should this be called `LernaRepository`? Maybe we support other mono repo types in the future?